### PR TITLE
Initial Scripture Burrito 0.2.0 schema

### LINKED ISSUE #70: Create First Attempt at JSON-Schema Representation

### PR DESCRIPTION
**Pull request by rdb**
This is a first pass at a JSON schema for Scripture Burrito 0.2.0. I'm publishing this a week in advance of the next working group meeting in the hope that people will have a chance to skim through at least the example document and the points of interest below.

For better organisation, the schema is split up into multiple files in the [schema folder](https://github.com/bible-technology/scripture-burrito/tree/0.2-json/schema), the main entry point being [metadata.schema.json](https://github.com/bible-technology/scripture-burrito/blob/0.2-json/schema/metadata.schema.json). The flavorDetails definitions are in flavorType-specific subdirectories.

For an example, see [the scriptureText.json example](https://github.com/bible-technology/scripture-burrito/blob/0.2-json/docs/examples/scriptureText.json).  To validate it:
```
pip install jsonschema
python code/validate.py docs/examples/scriptureText.json
```

### To Do
Most has been done except for the absence of publications and flavor details for flavors other than scriptureText and scriptureAudio. Due to limited availability on my end, I’m hoping others will be able to help with this.

- [ ] Define publication syntax.
- [ ] flavorDetails for flavors other than scriptureText or scriptureAudio shoul be validated.
- [ ] Add dateStarted/dateCompleted somewhere (see #89)
- [ ] Fill out more description strings.
- [ ] Create unit tests
- [ ] Integrate with documentation (see #71)

### Approach
I used the XML examples and Relax-NG schema as my guide. Since the aim here is not to provide something that could be roundtripped to a known XML representation, but to create a new primary representation for burritos, the priority should be on minimizing schema complexity and maximize machine processing utility. So while I've generally stuck to the existing convention and mapped those straightforwardly to JSON equivalents, there were a number of cases where I thought an alternative representation made more sense, subject to discussion (see notes below).

Note that although some programming languages (JS, Python 3.6+) have dictionary implementations that preserve insertion order, this is not a guaranteed property by JSON. Even in fields where order is not given any semantic relevance (eg. in the list of languages or countries), in cases where metadata is presented to a user, users may be liable to confusion if tools arbitrarily change the order of such fields, by just the process of parsing and writing. A solution to this is to mandate that user interfaces always display these fields in a particular order, but this may be unenforceable in practice. This proposal instead offers use arrays instead of dictionaries where there is no significantly added value from the use of a dictionary.

### Meta section
The burritoVersion needed a place. It made most sense to me to take the snapshot section and turn it into a “meta” section that essentially contains metadata of the metadata: information that should be read first which describes how to interpret the rest of the document. The information herein will likely not be retained by software that converts the Burrito to a different representation, and as such, it doesn’t contain information essential to interpreting the Burrito contents.

I’m not happy with the confusion between the name of “meta” and the concept of “metadata”; other suggestions welcome.

### Licenses
The licenses array has been moved to the copyright section (it seemed to make the most sense as this interacts with other aspects of copyright).

The rightsholders and rights administrator are now indices into the root agencies array. This simultaneously avoids duplication of agency info while allowing the preservation of the order in which rightsholders are presented.

https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/docs/examples/scriptureText.json#L127-L132

### Scope and Canon
bookScope is an associative array, to allow easy extraction of the books without having to parse the canonical reference syntax.  Validation is now also being done that no out-of-canon books are specified here.
canonSpec has been broken up into canonType and canonSpec, to allow both easy extraction of canon info and also for the canonType to be specified even if the detailed components are not known:
https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/docs/examples/scriptureText.json#L63-L67
Though, the canon syntax may need revision, especially as the validation ([one](https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/schema/type.schema.json#L101-L163), [two](https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/schema/type.schema.json#L232-L270)) may be more complex than it needs to be.

### Contributor roles are an array
Rather than a bunch of booleans, this is a more compact representation, and not really harder to work with (most systems will probably convert this to a set, internally).
https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/docs/examples/scriptureText.json#L109-L119

### systemId no longer prefixed
Identifiers are now specified without prefix in systemId, this seemed redundant.
https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/docs/examples/scriptureText.json#L27-L31

## Discussion points
There are some points that came up during the process that I think deserve some attention from the broader group. Should we create separate issues for these?

### Compatibility policy
We should have decided on a compatibility policy by the time we release 1.0.0, and set up a regression testing system to enforce it. Will a SB 1.0.1 file be valid against the 1.0.0 spec, and vice versa?
My schema includes a “minVersion” field that implementations have to check so that they can refuse, but this is pending based on the outcome of this discussion. Furthermore, “additionalProperties” is set to false in most places, indicating that fields unrecognized by the spec are disallowed rather than silently ignored, but this may also have to be revised depending on what we decide here.

### What is a minimal burrito?
What fields are truly required to be able to produce a minimal burrito? In my opinion, we should review which fields should be required, and produce a minimal burrito that just barely validates, that we can use as an automated test case.

### flavorDetails and custom fields
The structure and schema of flavorDetails differs from flavor to flavor. So, `type.flavorDetails.translationType` may mean completely different things depending on the value of `type.flavorDetails.flavor`. Is this considered unexpected to developers, or is this fine?
The alternative would be some syntax like the following:
```json
    "flavors": {
        "scriptureText": {
            "field": "value"
        }
    }
```
Based on that syntax, however, people might intuitively expect multi-flavor burritos to be supported:
```json
    "flavors": {
        "scriptureText": {...},
        "scriptureAudio": {...},
        "x-custom": {}
    }
```
That’s a direction we could choose to go into, but I’ll just leave this matter here for consideration. The current approach may be fine if we decide that it's not too unintuitive.

### flavorType/flavor redundancy
Currently, we have defined a flavorType and flavor:
```
    "flavor": "scriptureText",
    "flavorType": "scripture",
```
This seems to be redundant, requiring extra validation, because a flavor always belongs to a particular flavorType. We could consider instead having a single flavor field like so:
```
    "flavor": "scripture/text"
```
This seems to eliminate the redundancy while making it more obvious what the type is. (The syntax above evokes familiarity to MIME types, for better or worse; other syntaxes might be possible if this is undesirable.)

Custom flavors could still be specified like so:
```
    "flavor": "scripture/x-pieRecipe"
```

### flavor naming
Right now, the flavorType is *almost* always a prefix of the flavor: “scriptureText” is of type “scripture”, with the exception being “glossedTextStory” for the “gloss” type. This may make people likely to erroneously rely on this. I would recommend choosing between:

- Making this consistently so (the solution of “scripture/text” above would solve this)
- Making the flavor names so obviously non-derived from flavor types that people couldn’t be confused.

### XHTML dialect
SB adopts DBL's mini-dialect of XHTML for rich markup in the copyright and promotion statement fields. This is impossible to validate using JSON-schema. However, for security reasons, we still need to do some basic validation, eg. that no script tags are present that could enable malicious XSS attacks when naïve implementations display it in a browser.

I created these RegEx monstrosities which probably come as close as we can to exhaustive validation of this field, and should suffice for the security use-case and for validating most basic markup, but are not perfect:
https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/schema/common.schema.json#L64-L73

Do we want to continue using the XHTML dialect? Are there alternatives options, such as a JSON representation of the DOM?

### bookScope
Right now we have a custom language for ranges, allowing people to define things like:
```
    "bookScope": {
        "GEN": "1,2-3,4:4-6:7"
    }
```
This will require people to create a parser to extract the ranges. Should we try to instead create a JSON object representation for these ranges?

### scopeOrRole is borderline unparseable
The ingredient definition has a scopeOrRole field that can be pretty much anything and the validation rules include a regular expression longer than 1400 characters ([see schema](https://github.com/bible-technology/scripture-burrito/blob/0.2-json/schema/ingredient.schema.json#L21-L49)). I’m a little unfamiliar with the meaning of this field, but perhaps we should split this up into an object structure that makes it easier to interpret?

### Revise enum naming convention
We seem to be inconsistent on this matter. Especially under the flavorDetails definitions, where enums are occasionally defined with upper-case, sometimes as English strings, and othertimes as camelCase.  Also see #92. Should we adopt a consistent convention for this?

### Public domain
The publicDomain indicates that the burrito is placed into the public domain. Should rightsHolders be optional if this is set? Should it still be possible to set licenses if publicDomain is specified? I figure that in some jurisdictions that do not have a concept of public domain, it may be relevant to set “fallback licenses”, but IANAL.

Fixes #70

### CODE REVIEW

**Comment by rdb** [comment]
I've added Node.js validation scripts.

It is preferred in JSON schema to identify the schema using a canonical URL under which the schema can be accessed by validators.  What will this be?  I have set it to `https://burrito.bible/schema/*.schema.json` for now, but this can be changed.

**Comment by jag3773** [comment]
@rdb Can you rebase this feature branch off of `develop` instead of master?

**Comment by mvahowe** [comment]
Re the discussion points:

A metametadata section may turn out to be a good idea. It really depends how processors work, and to what extent we propose a processing model.

My main thought on the licenses proposal is that it makes the metadata very hard to read. That's ok if we provide some sort of metadata viewer. Otherwise, it's going to be a problem for archivists whenever they want to eyeball the metadata (which happens quite a lot right now.) Hand editing of the metadata becomes hasardous too. I would ask whether, in the future, we need to preserve order in this section. I know that some processing currently depends on it, but I don't think we should encourage that to remain the case with SB.

Scope and Canon looks sensible to me.

Contributor role enums are fine by me.

The argument for prefixing ids was that the ids then make sense beyond the metadata document. ie, if we display 55df02965117ad3f2201309b from the example, what does it mean? AFAICS we need to know which IDServer relates to and, in that case, we should consider a canonical way to provide that information. The uuid on its own is not at all portable.

My starting position on compatibility discussions is that there's no such thing as 100% backwards compatibility, because some processor out there may rely on the presence, absence, list of enums, etc of any obscure feature. The processor I'm working will allow explicit configuration of minimum and maximum schema version.

I chatted with @rdb today about minimal burritos. I think the place to start is what is currently required by the Registry, although we both think we'd eventually like to aim for more than that. I think we need, at the very least, a "start of project" schema and a "ready for publication" schema.

For me, the whole idea of the flavorDetails section is that it's specific to the flavor. Trying to avoid different flavors using the same key within flavorDetails in different ways seems to me to be a recipé for endless discussions with too many people in the room, because we suddenly need to consider every current and future flavor before adding any flavor-specific key.

I don't care much about flavortype/flavor syntax. In XML it's considered good practice not to require string operations to get basic information (ie not requiring myString.split("/")[0]). I also fear that adopting this syntax will recreate confusion between flavorType/flavor and mimetypes.

If we go for the / format, making the names consistent is fine by me. My scheme aimed at "easy to read outloud", which is a legitimate concern in XML, and which is presumably something no-one using JSON has ever considered :-)

I think we need some subset of XHTML for rich text. Multiple tools and widgets exist for editing this. If we roll our own format, everyone then needs to invent a wysiwyg editor for their users. Rigorous validation of XHTML via regexes is impossible, so I'd suggest having an alternative validation system for this, maybe a JS class which uses recursive DOM operations. (This is obviously one place where XML schema technology makes a lot more sense.)

bookScope and scopeOrRole does indeed assume a big and hairy regex. This regex is used in several places in the XML schema and also in other DBL applications. It's also close to what end users see in PT, and I think the string representation is going to be much much easier for end users to read than nested objects.

The convention enums are a mess, and also a charade since we currently have no definitions of conventions and no way to validate them if we did (see separate issue.)

I think most people would be confused by a public domain entry with a licence. (If we want to really confuse ourselves and everyone else, we should consider the KJV text which is public domain in the US but crown copyright in the UK.)

**Comment by mvahowe** [comment]
With the help of @rdb I have proof of concept validation running within my burrito stack project. I hope that this will make exploring the schema easier.

**Comment by jag3773** [comment]
@rdb I think I'll wait until you make some of the basic changes we discussed in the meeting today relative to your discussion points before merging this PR.  The items requiring more discussion we can handle in separate PRs.

**Comment by mvahowe** [comment]
On the prefixes for ids, again...

Say I have a burrito store that contains

- a burrito with id of 1234 in idServer abc

- a burrito with id of 1234 with idServer def

I have a catalog view that displays the id. How useful is "1234" as a unique identifier.

Or, I'm constructing a URL in a RESTful API. How useful is "1234" as part of that URL?

Or, I'm writing a SQL request. Doesn't the current proposal mean that I need a composite key?

Or, in general, if I come across 1234 somewhere, can I make any sense of it?

**Comment by rdb** [comment]
> I have a catalog view that displays the id. How useful is "1234" as a unique identifier.

Not sure I understand this point.  Also not sure why you would be displaying an identifier, especially without context, in a catalog view.  If you wanted to display identifiers for an entry, you would want to display pairs of system+id, which is straightforward to derive from systemId as-is.

> Or, I'm constructing a URL in a RESTful API. How useful is "1234" as part of that URL?

Not sure I understand this point either.  Isn't the idea that you will be able to query DBL for https://api.library.bible/entries/1234 as opposed to https://api.library.bible/entries/dbl::1234 (and certainly not https://api.library.bible/entries/uw::1234 ).

In fact, if I want to query DBL for this burrito, I expect I would actually need to go out of my way to strip the prefix to pass to the DBL API.  Not that that's a big deal at all, I'm just not sure what API you are imagining where the prefix would be passed in a URL.

> Or, I'm writing a SQL request. Doesn't the current proposal mean that I need a composite key?

You already need a composite key because the revision will be separate.  Depending on which database system you use, you might actually be using one index per system, in which case having the ids prefixed does not help anyway.

> Or, in general, if I come across 1234 somewhere, can I make any sense of it?

Isn't that true for any value in the document?  If you come across "1,3-12", or "public", or "0.1" somewhere, can you make sense of those?  Or even "dbl::23"; how do you know whether it's an organisation id, user id, or entry id?

**Comment by mvahowe** [comment]
@rdb I think you may have missed the start of my example, where a single burrito store may contain burritos with ids from many idServers and, also, where the burrito store may not have its own id for some burritos that it stores. That is a likely scenario for an editor, for example, where, eg, the texts being edited are from uW but reference texts ("PT resources") may come from DBL or Wycliffe Associates. In that context, the id on its own is not a useful key.

**Comment by mvahowe** [comment]
> You already need a composite key because the revision will be separate.

There are actually no views on the DBL website right now that show you a specitic entry/revision. An entry makes sense in its own right - that's surely what the Registry does, and the only basis on which it is possible to report progress. In general, the semantics of an entry view are "the latest revision of this entry".

However, even now, we need to know whether someone is referring to a PT project or a DBL entry. In some future world the project id might be shared between PT and DBL but, even in this case, we need to know if it's ptx::myproject or dbl::myproject because the PT project will often be ahead of the DBL entry.

**Comment by mvahowe** [comment]
> The rightsholders and rights administrator are now indices into the root agencies array.

Having thought about this some more, I think this is a step too far towards optimising the JSON as an internal data representation.

```"rightsHolderAgencies": [0, 1]```

is hard to read but, more importantly, it's fragile. If the list of agencies is edited, and this indexing is not updated accordingly, we're screwed. All processors would need something like DB transactions to ensure that this didn't happen.

Part of the motivation for this is preserving order. But, in SB, the order here (and in general) does not have any semantics. If pre-SB processors do rely on order, I think that's a migration issue.

I think that a list of ids would do the job at least as well, with almost no bloat, ie

```"rightsHolderAgencies": ["dbl::123", "agmt::123"]```

**Comment by jag3773** [comment]
https://github.com/bible-technology/scripture-burrito/blob/7203a081320559f723d32e2265056b1c2b739c03/docs/examples/scriptureText.json#L69-L79

When I saw this relationships section I changed my mind on the prefixing in the systemID section.  Since the ids will be used elsewhere in the document, outside of the systemID context, I *do not think* it is redundant to specify them in the systemID section.

In other words, I agree with @mvahowe that it would be good to always represent the systemID prefix.  I think this is more consistent and will make it easier to reference without needing to parse the field in some circumstances but not others.

**Comment by rdb** [comment]
It sounds like some use cases will require stripping the prefix and others will require leaving it on.

My main concern is that it's impossible to validate in the schema that the prefix is the same as the dictionary key.

> `"rightsHolderAgencies": ["dbl::123", "agmt::123"]`

I'm fine with this.

**Comment by mvahowe** [comment]
> My main concern is that it's impossible to validate in the schema that the prefix is the same as the dictionary key.

How bad would it be to have a list of ids? Then there's no key to validate. It takes a few more clock cycles to find the id you want, but it's not like we're going to have 1000 of them.

**Comment by mvahowe** [comment]
@rdb @jag3773 I'm writing code that uses this schema and spotting some issues. But I think that it would be easier to tackle them one issue at a time. So maybe make the agreed changes, merge, and we'll take it from there?

**Comment by rdb** [comment]
I forgot, did we agree on adding the schema URI and then keeping the version separately, or should we parse the version out of the schema URI?

> Oh, and the other reason is that we **have to do this**! IDs are only guaranteed to be unique within the context of a server.  So it's entirely possible that `gl47` is a valid id for unfoldingWord and `x-atl`, without the prefix in the ID it's ambiguous which one is being referenced.

I get the consistency argument, but not the ambiguity one, because you will never have a systemId that goes `"x-atl": "uw::gl47"` (but we can't validate that).

> How bad would it be to have a list of ids? Then there's no key to validate. It takes a few more clock cycles to find the id you want, but it's not like we're going to have 1000 of them.

I'm not opposed to this—this actually may make indexing easier in MongoDB—but it's now impossible to validate that the same prefix doesn't occur twice, so it doesn't solve the validation problem.

There's a proposal for [key-based item uniqueness](https://github.com/json-schema-org/json-schema-spec/issues/538) which, if accepted to JSON-Schema, would allow `[{"system": "dbl", "id": "abcd"}]`.

**Comment by mvahowe** [comment]
Today's WG meeting agreed to merge this PR and use separate PRs and issues to discuss outstanding details. Thank-you, @rdb, for your work on this!
